src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocSpanExpressionsTests.cs

```diff
@@ -0,0 +1,495 @@
+            Assert.Equal(4, variables.Count());
+
+            var obj1 = variables.ElementAt(0);
+            Assert.Equal("obj1", obj1.Identifier.Text);
+            Assert.Equal(SyntaxKind.CastExpression, obj1.Initializer.Value.Kind());
+
+            var obj1Value = model.GetSemanticInfoSummary(((CastExpressionSyntax)obj1.Initializer.Value).Expression);
+            Assert.Null(obj1Value.Type);
+            Assert.Equal("Span", obj1Value.ConvertedType.Name);
+            Assert.Equal(ConversionKind.StackAllocToSpanType, obj1Value.ImplicitConversion.Kind);
+            Assert.Equal(ConversionKind.Identity, obj1Value.ImplicitConversion.UnderlyingConversions.Single().Kind);
+
+            var obj2 = variables.ElementAt(1);
+            Assert.Equal("obj2", obj2.Identifier.Text);
+
+            var obj2Value = model.GetSemanticInfoSummary(obj2.Initializer.Value);
+            Assert.Null(obj2Value.Type);
+            Assert.True(obj2Value.ConvertedType is PointerTypeSymbol);
+            Assert.Equal("Int32", ((PointerTypeSymbol)obj2Value.ConvertedType).PointedAtType.Name);
+            Assert.Equal(ConversionKind.StackAllocToPointerType, obj2Value.ImplicitConversion.Kind);
+            Assert.Equal(ConversionKind.Identity, obj2Value.ImplicitConversion.UnderlyingConversions.Single().Kind);
+
+            var obj3 = variables.ElementAt(2);
+            Assert.Equal("obj3", obj3.Identifier.Text);
+
+            var obj3Value = model.GetSemanticInfoSummary(obj3.Initializer.Value);
+            Assert.Null(obj3Value.Type);
+            Assert.Equal("Span", obj3Value.ConvertedType.Name);
+            Assert.Equal(ConversionKind.StackAllocToSpanType, obj3Value.ImplicitConversion.Kind);
+            Assert.Equal(ConversionKind.Identity, obj3Value.ImplicitConversion.UnderlyingConversions.Single().Kind);
+
+            var obj4 = variables.ElementAt(3);
+            Assert.Equal("obj4", obj4.Identifier.Text);
+
+            var obj4Value = model.GetSemanticInfoSummary(obj4.Initializer.Value);
+            Assert.Null(obj4Value.Type);
+            Assert.True(obj4Value.ConvertedType is PointerTypeSymbol);
+            Assert.Equal("Int32", ((PointerTypeSymbol)obj4Value.ConvertedType).PointedAtType.Name);
+            Assert.Equal(ConversionKind.StackAllocToPointerType, obj4Value.ImplicitConversion.Kind);
+            Assert.Equal(ConversionKind.Identity, obj4Value.ImplicitConversion.UnderlyingConversions.Single().Kind);
+        }
+
+        [Fact]
+        public void ConversionError()
+        {
+            CreateCompilationWithMscorlibAndSpan(@"
+class Test
+{
+    void M()
+    {
+        double x = stackalloc int[10];          // implicit
+        short y = (short)stackalloc int[10];    // explicit
+    }
+}", TestOptions.UnsafeReleaseDll).VerifyDiagnostics(
+                // (6,20): error CS8520: Conversion of a stackalloc expression of type 'int' to type 'double' is not possible.
```

Comment:
I think conversion operators on Span type are irrelvant here since the stackalloc expression has no type and we look at conversions "from expression" which would only allow converting to span or a pointer.

The error that @gafter proposes indeed sounds clearer though.